clippy_lints/src/str_to_string.rs

```diff
@@ -0,0 +1,55 @@
+use crate::utils::is_type_diagnostic_item;
+use crate::utils::span_lint_and_help;
+
+use if_chain::if_chain;
+use rustc_hir as hir;
+use rustc_hir::*;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** This lint checks for `.to_string()` method calls on values of type `&str`.
+    ///
+    /// **Why is this bad?** This does a multitude of things to just clone a string. Using `.to_owned()` is more specific.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// // example code where clippy issues a warning
+    /// let _ = "str".to_string()
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// // example code which does not raise clippy warning
+    /// let _ = "str".to_owned()
+    /// ```
+    pub STR_TO_STRING,
+    restriction,
+    "using `to_string()` on a str, which should be `to_owned()`"
+}
+
+declare_lint_pass!(StrToString => [STR_TO_STRING]);
+
+impl LateLintPass<'_> for StrToString {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &hir::Expr<'_>) {
+        if_chain! {
+            if let ExprKind::MethodCall(path, _, _args, _) = &expr.kind;
+            if path.ident.name == sym!(to_string);
+            then {
+                let ty = cx.typeck_results().expr_ty(expr);
+                if is_type_diagnostic_item(cx, ty, sym!(string_type)) {
```

Comment:
Ah, you have to check the type of `args[0]`, not the type of `expr`. The type of `expr` will ever be `String`, since `to_string` returns a `String`.

Reply:
Yes, works now. Thank you